### PR TITLE
Add mutate in evolutionary search

### PR DESCRIPTION
**Pull request by BiynXu**
1.Add the mutation process of population in EvolutionarySearch. 
Since the mutation of each SearchState is independent and can be parallelized, multiple threads are started, and each thread independently performs a state mutation.
2.Add a unit test of Evolve() and modify some original unit tests have been modified.

### CODE REVIEW

File: cinn/auto_schedule/search_strategy/evolutionary_search.cc [hunk 1]
```diff
@@ -145,10 +194,29 @@ std::vector<SearchState> EvolutionarySearch::Evolve(const std::vector<SearchStat
     }
     evolution.push_back(CrossOver(population[first_rand_idx], population[second_rand_idx]));
   }
-
+  // mutate
+  std::vector<SearchState> mutated_individuals(evolution.size());
```
Comment (CtfGo) [inline cinn/auto_schedule/search_strategy/evolutionary_search.cc:201]: 加点VLOG日志区分Evolve中各阶段的运行

File: cinn/auto_schedule/cost_model/expr_cost_model.h [hunk 2]
```diff
@@ -29,7 +29,7 @@ namespace auto_schedule {
  */
 class ExprCostModel : public XgbCostModel {
  public:
-  float Predict(const ir::ModuleExpr& sample, const common::Target& target) const;
+  virtual float Predict(const ir::ModuleExpr& sample, const common::Target& target) const;
```
Comment (CtfGo) [inline cinn/auto_schedule/cost_model/expr_cost_model.h:32]: 后续看下这个Predict升级支持批量预估是否可行

File: cinn/auto_schedule/search_strategy/evolutionary_search.h [hunk 3]
```diff
@@ -42,7 +43,8 @@ class EvolutionarySearch {
   EvolutionarySearch(const TuneTask& tune_task,
                      const ExprCostModel& cost_model,
                      Database* database,
-                     utils::LinearRandomEngine::StateType rand_seed = -1);
+                     utils::LinearRandomEngine::StateType rand_seed                   = -1,
+                     const std::vector<std::tuple<std::string, double>>& mutate_rules = {});
```
Comment (CtfGo) [inline cinn/auto_schedule/search_strategy/evolutionary_search.h:47]: TaskOptimizer调用处还未升级，是下个pr再提还是有什么考虑

File: cinn/auto_schedule/cost_model/expr_cost_model.h (hunk 2 above)
Comment (BiynXu) [reply cinn/auto_schedule/cost_model/expr_cost_model.h:32]: 后续再提一个Pr来支持批量Predict

File: cinn/auto_schedule/search_strategy/evolutionary_search.h (hunk 3 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_strategy/evolutionary_search.h:47]: 目前只有一个mutate rule，使用默认权重1。未来调该参数时再升级TaskOptimizer和TuningOptions。

**Review by CtfGo — APPROVED** [review]

**Pull request merged**